### PR TITLE
Include the query string parameters when redirecting to a new URL

### LINKED ISSUE #3061: query parameters dropped when redirecting from hub to login page
### Bug description
query parameters are dropped when redirecting from hub to login page.   Documentation which I list below says query parameters should be preserved.

From hub log you can see the parameters are dropped on redirect:
[I 2020-05-29 05:55:28.094 JupyterHub log:174] 302 GET /hub/?frog=water -> /hub/login (@10.0.1.5) 1.01ms
[I 2020-05-29 05:55:28.147 JupyterHub log:174] 302 GET /hub/login -> /hub/oauth_login (@10.0.1.5) 0.91ms

Documentation states:
https://jupyterhub.readthedocs.io/en/stable/reference/urls.html
All authenticated handlers redirect to /hub/login to login users prior to being redirected back to the originating page. The returned request should preserve all query parameters.

#### Expected behaviour
query parameters would be preserved on redirects

#### Actual behaviour
query parameters are dropped

### How to reproduce

1. Go to /hub?x=y
2. It will redirect to /login and your url parameters will be gone

### Your personal set up

 - OS:
linux
 - Version:
JupyterHub version 1.1.0
 - Configuration:

### PR DESCRIPTION
**Pull request by kinow**
Closes #3061

Tentative for #3061, will re-read the issue and my own changes, and then prepare tests if it's looking good.

I changed recently some code in another PR that was handling parameters in URL redirect, but with JavaScript. When I read the issue #3061, I thought I could have caused a regression, and decided to investigate it.

#3061 is not related to my change. But I thought the bug reported made sense, and keeping the parameters when redirecting could be useful (e.g. when passing some token/sessionId/flag/etc down to a spawner).

Cheers
Bruno

### CODE REVIEW

**Comment by kinow** [comment]
Tests passing. Just need to write some extra tests, and then ready for review.

File: jupyterhub/handlers/base.py [hunk 1]
```diff
@@ -654,8 +654,36 @@ def get_next_url(self, user=None, default=None):
                     next_url = url_path_join(self.hub.base_url, 'spawn')
             else:
                 next_url = url_path_join(self.hub.base_url, 'home')
+
+        next_url = self.append_query_parameters(next_url, exclude=['next'])
         return next_url
 
+    def append_query_parameters(self, url, exclude=None):
```
Comment (minrk) [inline jupyterhub/handlers/base.py:661]: Maybe the default should be to exclude `next` since it's called that way 100% of the time so far?

Comment (kinow) [reply jupyterhub/handlers/base.py:661]: Hmmm, good point. That would prevent other developers using it and accidentally including `next`, which could trigger a redirect-loop. Let me update it.

**Comment by minrk** [comment]
Nice! Makes sense. Left one comment in-line about defaults matching common (universal?) usage, but looks great to me.

**Comment by kinow** [comment]
> Nice! Makes sense. Left one comment in-line about defaults matching common (universal?) usage, but looks great to me.

Thanks for the review @minrk . I've added a [new commit](https://github.com/jupyterhub/jupyterhub/pull/3089/commits/83eaa028dec74eaa15c03f1bab9c90a201ab864b) with a tentative pydocs text and default `exclude=["none"]` (happy to squash it later if necessary).

I did that within the function, instead of in the signature to avoid the mutable parameter, but happy to update it if there's another way that works better for this.

**Comment by kinow** [comment]
One unit test failure on 3.5. Tried to reproduce but it passes for me (3.7-anaconda). It looks
like I wrote a fragile unit test. So instead of checking the exact string, will test simply that it
contains both parameters in the string, as from the log output in CI it appears the order
the parameters are added is not guaranteed.

```py
assert r.url.endswith('/hub/login?next=foo&param1=test')

assert 'next=foo' in r.url
assert 'param1=test' in r.url
```

Edited last commit with the changes above. Let's see if CI builds pass now.

File: jupyterhub/handlers/login.py [hunk 2]
```diff
@@ -90,13 +90,20 @@ class LoginHandler(BaseHandler):
     """Render the login page."""
 
     def _render(self, login_error=None, username=None):
+        # login url with any parameters specified by the user/application
+        login_url = self.append_query_parameters(
+            self.settings['login_url'], exclude=['next']
+        )
+        # the next url for after the login
+        next_url = url_escape(self.get_argument('next', default=''))
```
Comment (minrk) [inline jupyterhub/handlers/login.py:98]: I don't think it's quite right that the URL parameters from the login page are preserved on the destination. Instead, they should be *encoded* in the `next` URL. Adding them here might be doubly-encoding params.

File: jupyterhub/handlers/base.py [hunk 3]
```diff
@@ -654,8 +654,42 @@ def get_next_url(self, user=None, default=None):
                     next_url = url_path_join(self.hub.base_url, 'spawn')
             else:
                 next_url = url_path_join(self.hub.base_url, 'home')
+
+        next_url = self.append_query_parameters(next_url, exclude=['next'])
```
Comment (minrk) [inline jupyterhub/handlers/base.py:658]: I think this addition of params is the one that's definitely correct, adding params where `next` is already defined is probably not correct.

File: jupyterhub/handlers/login.py (hunk 2 above)
Comment (kinow) [reply jupyterhub/handlers/login.py:98]: Ah! I think that makes sense! Let me try that.

Comment (kinow) [reply jupyterhub/handlers/login.py]: Had a try at adding the tests and adding the parameters escaped/encoded to the `next` parameter @minrk . Let me know if that looks better :+1:

File: jupyterhub/tests/test_pages.py [hunk 4]
```diff
@@ -516,6 +517,58 @@ async def test_user_redirect_deprecated(app, username):
...
+        (
+            # login
+            # simplest case, accessing the login URL, gives an empty next url
+            'login',
+            [],
+            '/hub/login',
+            '/hub/login?next=',
+        ),
+    ],
+)
```
Comment (kinow) [inline jupyterhub/tests/test_pages.py:557]: Realized today that this test is quite similar to `test_login_redirect`, with the difference that it follows redirects, matching with the final URL, and also checks the `<form action="">` value. I wonder whether that's OK, or if it needs to be merged/updated to fit into `test_login_redirect` or renamed, etc?

Comment (minrk) [reply jupyterhub/tests/test_pages.py:557]: I think it's fine to add this to focus on slightly different things.

File: jupyterhub/handlers/base.py [hunk 5]
```diff
@@ -654,8 +654,42 @@ def get_next_url(self, user=None, default=None):
...
+        as most use cases in JupyterHub (all?) won't want to include the next
+        parameter twice (the next parameter is added elsewhere to the query
+        parameters).
+
+        :param str url: a URL
+        :param list exclude: optional list of parameters to be ignored, defaults to
+        a list with "next" (to avoid redirect-loops)
+        :rtype (str)
+        """
+        if not exclude:
```
Comment (minrk) [inline jupyterhub/handlers/base.py:681]: Checking `if not exclude` prevents callers from specifying `exclude=[]`. Explicit None check will more accurately capture the default:

```suggestion
        if exclude is None:
```

Comment (kinow) [reply jupyterhub/handlers/base.py:681]: Oops! Fixing it!

**Comment by minrk** [comment]
Thanks!

**Comment by kinow** [comment]
Thank you for the review @minrk !

**Pull request merged**